File: go/ingest/hooks/mutation_hook_test.go
// SPDX-License-Identifier: Apache-2.0
package hooks

import (
	"sync"
	"sync/atomic"
	"testing"
	"time"

	"github.com/jeffs-brain/memory/go/brain"
)

func TestCreatedEventInRawDocumentsDispatches(t *testing.T) {
	var dispatched atomic.Int32
	var dispatchedPath string
	var mu sync.Mutex

	hook := NewMutationHook(MutationHookOptions{
		BrainID:          "brain-1",
		DebounceInterval: 10 * time.Millisecond,
		Dispatch: func(brainID, path string) error {
			mu.Lock()
			dispatchedPath = path
			mu.Unlock()
			dispatched.Add(1)
			return nil
		},
	})
	defer hook.Close()

	sink := hook.Sink()
	sink.OnBrainChange(brain.ChangeEvent{
		Kind: brain.ChangeCreated,
		Path: "raw/documents/readme.md",
		When: time.Now(),
	})

	time.Sleep(50 * time.Millisecond)

	if got := dispatched.Load(); got != 1 {
		t.Fatalf("expected 1 dispatch, got %d", got)
	}
	mu.Lock()
	if dispatchedPath != "raw/documents/readme.md" {
		t.Fatalf("expected raw/documents/readme.md, got %q", dispatchedPath)
	}
	mu.Unlock()
}

func TestUpdatedEventInRawDocumentsDispatches(t *testing.T) {
	var dispatched atomic.Int32

	hook := NewMutationHook(MutationHookOptions{
		BrainID:          "brain-1",
		DebounceInterval: 10 * time.Millisecond,
		Dispatch: func(_, _ string) error {
			dispatched.Add(1)
			return nil
		},
	})
	defer hook.Close()

	sink := hook.Sink()
	sink.OnBrainChange(brain.ChangeEvent{
		Kind: brain.ChangeUpdated,
		Path: "raw/documents/notes.md",
		When: time.Now(),
	})

	time.Sleep(50 * time.Millisecond)

	if got := dispatched.Load(); got != 1 {
		t.Fatalf("expected 1 dispatch, got %d", got)
	}
}

func TestDeletedEventIgnored(t *testing.T) {
	var dispatched atomic.Int32

	hook := NewMutationHook(MutationHookOptions{
		BrainID:          "brain-1",
		DebounceInterval: 10 * time.Millisecond,
		Dispatch: func(_, _ string) error {
			dispatched.Add(1)
			return nil
		},
	})
	defer hook.Close()

	sink := hook.Sink()
	sink.OnBrainChange(brain.ChangeEvent{
		Kind: brain.ChangeDeleted,
		Path: "raw/documents/readme.md",
		When: time.Now(),
	})

	time.Sleep(50 * time.Millisecond)

	if got := dispatched.Load(); got != 0 {
		t.Fatalf("expected 0 dispatches for delete, got %d", got)
	}
}

func TestPathOutsideRawDocumentsIgnored(t *testing.T) {
	var dispatched atomic.Int32

	hook := NewMutationHook(MutationHookOptions{
		BrainID:          "brain-1",
		DebounceInterval: 10 * time.Millisecond,
		Dispatch: func(_, _ string) error {
			dispatched.Add(1)
			return nil
		},
	})
	defer hook.Close()

	sink := hook.Sink()
	sink.OnBrainChange(brain.ChangeEvent{
		Kind: brain.ChangeCreated,
		Path: "memory/global/fact.md",
		When: time.Now(),
	})

	time.Sleep(50 * time.Millisecond)

	if got := dispatched.Load(); got != 0 {
		t.Fatalf("expected 0 dispatches for non-matching path, got %d", got)
	}
}

func TestCustomPathMatcher(t *testing.T) {
	var dispatched atomic.Int32

	hook := NewMutationHook(MutationHookOptions{
		BrainID:          "brain-1",
		DebounceInterval: 10 * time.Millisecond,
		PathMatchers:     []PathMatcher{PrefixPathMatcher("custom/")},
		Dispatch: func(_, _ string) error {
			dispatched.Add(1)
			return nil
		},
	})
	defer hook.Close()

	sink := hook.Sink()
	sink.OnBrainChange(brain.ChangeEvent{
		Kind: brain.ChangeCreated,
		Path: "custom/data.json",
		When: time.Now(),
	})

	time.Sleep(50 * time.Millisecond)

	if got := dispatched.Load(); got != 1 {
		t.Fatalf("expected 1 dispatch for custom path, got %d", got)
	}
}

func TestDebounceCoalescesRapidWrites(t *testing.T) {
	var dispatched atomic.Int32

	hook := NewMutationHook(MutationHookOptions{
		BrainID:          "brain-1",
		DebounceInterval: 100 * time.Millisecond,
		Dispatch: func(_, _ string) error {
			dispatched.Add(1)
			return nil
		},
	})
	defer hook.Close()

	sink := hook.Sink()
	// Rapid writes to the same path within debounce window.
	for range 5 {
		sink.OnBrainChange(brain.ChangeEvent{
			Kind: brain.ChangeUpdated,
			Path: "raw/documents/rapid.md",
			When: time.Now(),
		})
		time.Sleep(10 * time.Millisecond)
	}

	time.Sleep(200 * time.Millisecond)

	if got := dispatched.Load(); got != 1 {
		t.Fatalf("expected 1 coalesced dispatch, got %d", got)
	}
}

func TestOptOutByBatchReason(t *testing.T) {
	var dispatched atomic.Int32

	hook := NewMutationHook(MutationHookOptions{
		BrainID:          "brain-1",
		DebounceInterval: 10 * time.Millisecond,
		OptOutReasons:    map[string]bool{"pipeline": true, "ingest": true},
		Dispatch: func(_, _ string) error {
			dispatched.Add(1)
			return nil
		},
	})
	defer hook.Close()

	sink := hook.Sink()

	// Event with opt-out reason should be ignored.
	sink.OnBrainChange(brain.ChangeEvent{
		Kind:   brain.ChangeCreated,
		Path:   "raw/documents/pipeline-output.md",
		Reason: "pipeline",
		When:   time.Now(),
	})

	time.Sleep(50 * time.Millisecond)

	if got := dispatched.Load(); got != 0 {
		t.Fatalf("expected 0 dispatches for opt-out reason, got %d", got)
	}

	// Event without opt-out reason should dispatch.
	sink.OnBrainChange(brain.ChangeEvent{
		Kind:   brain.ChangeCreated,
		Path:   "raw/documents/user-upload.md",
		Reason: "user-write",
		When:   time.Now(),
	})

	time.Sleep(50 * time.Millisecond)

	if got := dispatched.Load(); got != 1 {
		t.Fatalf("expected 1 dispatch for non-opt-out reason, got %d", got)
	}
}

func TestDifferentPathsDispatchIndependently(t *testing.T) {
	var dispatched atomic.Int32

	hook := NewMutationHook(MutationHookOptions{
		BrainID:          "brain-1",
		DebounceInterval: 10 * time.Millisecond,
		Dispatch: func(_, _ string) error {
			dispatched.Add(1)
			return nil
		},
	})
	defer hook.Close()

	sink := hook.Sink()
	sink.OnBrainChange(brain.ChangeEvent{
		Kind: brain.ChangeCreated,
		Path: "raw/documents/a.md",
		When: time.Now(),
	})
	sink.OnBrainChange(brain.ChangeEvent{
		Kind: brain.ChangeCreated,
		Path: "raw/documents/b.md",
		When: time.Now(),
	})

	time.Sleep(50 * time.Millisecond)

	if got := dispatched.Load(); got != 2 {
		t.Fatalf("expected 2 dispatches for different paths, got %d", got)
	}
}

func TestGlobPathMatcher(t *testing.T) {
	matcher := GlobPathMatcher("raw/documents/**/*.md")

	tests := []struct {
		path  string
		match bool
	}{
		{"raw/documents/readme.md", true},
		{"raw/documents/sub/notes.md", true},
		{"raw/documents/sub/deep/nested.md", true},
		{"raw/documents/readme.txt", false},
		{"memory/global/fact.md", false},
	}

	for _, tt := range tests {
		if got := matcher(tt.path); got != tt.match {
			t.Errorf("glob match %q: expected %v, got %v", tt.path, tt.match, got)
		}
	}
}

func TestCloseStopsPendingTimers(t *testing.T) {
	var dispatched atomic.Int32

	hook := NewMutationHook(MutationHookOptions{
		BrainID:          "brain-1",
		DebounceInterval: 200 * time.Millisecond,
		Dispatch: func(_, _ string) error {
			dispatched.Add(1)
			return nil
		},
	})

	sink := hook.Sink()
	sink.OnBrainChange(brain.ChangeEvent{
		Kind: brain.ChangeCreated,
		Path: "raw/documents/closing.md",
		When: time.Now(),
	})

	// Close before debounce fires.
	hook.Close()
	time.Sleep(300 * time.Millisecond)

	if got := dispatched.Load(); got != 0 {
		t.Fatalf("expected 0 dispatches after close, got %d", got)
	}
}

File: sdks/ts/memory/src/ingest/hooks/mutation-hook.test.ts
// SPDX-License-Identifier: Apache-2.0

import { describe, expect, it, vi } from 'vitest'
import type { ChangeEvent } from '../../store/index.js'
import type { Path } from '../../store/index.js'
import {
  createMutationHook,
  defaultPathMatcher,
  globPathMatcher,
  prefixPathMatcher,
} from './mutation-hook.js'

const makeEvent = (
  kind: ChangeEvent['kind'],
  path: string,
  reason?: string,
): ChangeEvent => ({
  kind,
  path: path as Path,
  reason,
  when: new Date(),
})

describe('mutation-hook', () => {
  it('dispatches on created event in raw/documents/', async () => {
    const dispatch = vi.fn()
    const hook = createMutationHook({
      brainId: 'brain-1',
      dispatch,
      debounceIntervalMs: 10,
    })

    hook.sink(makeEvent('created', 'raw/documents/readme.md'))
    await delay(50)

    expect(dispatch).toHaveBeenCalledWith('brain-1', 'raw/documents/readme.md')
    hook.close()
  })

  it('dispatches on updated event in raw/documents/', async () => {
    const dispatch = vi.fn()
    const hook = createMutationHook({
      brainId: 'brain-1',
      dispatch,
      debounceIntervalMs: 10,
    })

    hook.sink(makeEvent('updated', 'raw/documents/notes.md'))
    await delay(50)

    expect(dispatch).toHaveBeenCalledTimes(1)
    hook.close()
  })

  it('ignores deleted events', async () => {
    const dispatch = vi.fn()
    const hook = createMutationHook({
      brainId: 'brain-1',
      dispatch,
      debounceIntervalMs: 10,
    })

    hook.sink(makeEvent('deleted', 'raw/documents/readme.md'))
    await delay(50)

    expect(dispatch).not.toHaveBeenCalled()
    hook.close()
  })

  it('ignores paths outside default pattern', async () => {
    const dispatch = vi.fn()
    const hook = createMutationHook({
      brainId: 'brain-1',
      dispatch,
      debounceIntervalMs: 10,
    })

    hook.sink(makeEvent('created', 'memory/global/fact.md'))
    await delay(50)

    expect(dispatch).not.toHaveBeenCalled()
    hook.close()
  })

  it('supports custom path matchers', async () => {
    const dispatch = vi.fn()
    const hook = createMutationHook({
      brainId: 'brain-1',
      dispatch,
      pathMatchers: [prefixPathMatcher('custom/')],
      debounceIntervalMs: 10,
    })

    hook.sink(makeEvent('created', 'custom/data.json'))
    await delay(50)

    expect(dispatch).toHaveBeenCalledWith('brain-1', 'custom/data.json')
    hook.close()
  })

  it('debounces rapid writes to the same path', async () => {
    const dispatch = vi.fn()
    const hook = createMutationHook({
      brainId: 'brain-1',
      dispatch,
      debounceIntervalMs: 100,
    })

    for (let i = 0; i < 5; i++) {
      hook.sink(makeEvent('updated', 'raw/documents/rapid.md'))
      await delay(10)
    }

    await delay(200)

    expect(dispatch).toHaveBeenCalledTimes(1)
    hook.close()
  })

  it('dispatches independently for different paths', async () => {
    const dispatch = vi.fn()
    const hook = createMutationHook({
      brainId: 'brain-1',
      dispatch,
      debounceIntervalMs: 10,
    })

    hook.sink(makeEvent('created', 'raw/documents/a.md'))
    hook.sink(makeEvent('created', 'raw/documents/b.md'))
    await delay(50)

    expect(dispatch).toHaveBeenCalledTimes(2)
    hook.close()
  })

  it('opts out by batch reason', async () => {
    const dispatch = vi.fn()
    const hook = createMutationHook({
      brainId: 'brain-1',
      dispatch,
      optOutReasons: new Set(['pipeline', 'ingest']),
      debounceIntervalMs: 10,
    })

    hook.sink(makeEvent('created', 'raw/documents/pipeline-output.md', 'pipeline'))
    await delay(50)

    expect(dispatch).not.toHaveBeenCalled()

    // Non-opt-out reason should dispatch.
    hook.sink(makeEvent('created', 'raw/documents/user-upload.md', 'user-write'))
    await delay(50)

    expect(dispatch).toHaveBeenCalledTimes(1)
    hook.close()
  })

  it('close() stops pending timers', async () => {
    const dispatch = vi.fn()
    const hook = createMutationHook({
      brainId: 'brain-1',
      dispatch,
      debounceIntervalMs: 200,
    })

    hook.sink(makeEvent('created', 'raw/documents/closing.md'))
    hook.close()

    await delay(300)
    expect(dispatch).not.toHaveBeenCalled()
  })

  it('events after close are ignored', async () => {
    const dispatch = vi.fn()
    const hook = createMutationHook({
      brainId: 'brain-1',
      dispatch,
      debounceIntervalMs: 10,
    })

    hook.close()
    hook.sink(makeEvent('created', 'raw/documents/late.md'))
    await delay(50)

    expect(dispatch).not.toHaveBeenCalled()
  })

  it('rejects paths containing ".."', async () => {
    const dispatch = vi.fn()
    const warnFn = vi.fn()
    const hook = createMutationHook({
      brainId: 'brain-1',
      dispatch,
      debounceIntervalMs: 10,
      logger: { debug: vi.fn(), info: vi.fn(), warn: warnFn, error: vi.fn() },
    })

    hook.sink(makeEvent('created', 'raw/documents/../../../etc/passwd'))
    hook.sink(makeEvent('created', 'raw/documents/..secret.md'))
    hook.sink(makeEvent('created', '../raw/documents/readme.md'))
    await delay(50)

    expect(dispatch).not.toHaveBeenCalled()
    expect(warnFn).toHaveBeenCalledTimes(3)
    hook.close()
  })

  it('ignores empty path', async () => {
    const dispatch = vi.fn()
    const hook = createMutationHook({
      brainId: 'brain-1',
      dispatch,
      debounceIntervalMs: 10,
    })

    hook.sink(makeEvent('created', ''))
    await delay(50)

    expect(dispatch).not.toHaveBeenCalled()
    hook.close()
  })

  it('dispatch errors are logged', async () => {
    const errorFn = vi.fn()
    const hook = createMutationHook({
      brainId: 'brain-1',
      dispatch: () => { throw new Error('dispatch fail') },
      debounceIntervalMs: 10,
      logger: { debug: vi.fn(), info: vi.fn(), warn: vi.fn(), error: errorFn },
    })

    hook.sink(makeEvent('created', 'raw/documents/failing.md'))
    await delay(50)

    expect(errorFn).toHaveBeenCalled()
    hook.close()
  })
})

describe('defaultPathMatcher', () => {
  it('matches raw/documents/ paths', () => {
    expect(defaultPathMatcher('raw/documents/readme.md')).toBe(true)
    expect(defaultPathMatcher('raw/documents/sub/deep.md')).toBe(true)
  })

  it('rejects non-matching paths', () => {
    expect(defaultPathMatcher('memory/global/fact.md')).toBe(false)
    expect(defaultPathMatcher('raw/documents/')).toBe(false)
  })

  it('rejects empty path', () => {
    expect(defaultPathMatcher('')).toBe(false)
  })
})

describe('globPathMatcher', () => {
  const matcher = globPathMatcher('raw/documents/**/*.md')

  it('matches markdown files under raw/documents/', () => {
    expect(matcher('raw/documents/readme.md')).toBe(true)
    expect(matcher('raw/documents/sub/notes.md')).toBe(true)
    expect(matcher('raw/documents/sub/deep/nested.md')).toBe(true)
  })

  it('rejects non-matching files', () => {
    expect(matcher('raw/documents/readme.txt')).toBe(false)
    expect(matcher('memory/global/fact.md')).toBe(false)
  })

  it('matches ** at end of glob', () => {
    const allMatcher = globPathMatcher('raw/documents/**')
    expect(allMatcher('raw/documents/readme.md')).toBe(true)
    expect(allMatcher('raw/documents/sub/notes.md')).toBe(true)
    expect(allMatcher('memory/global/fact.md')).toBe(false)
  })
})

describe('prefixPathMatcher boundary', () => {
  const matcher = prefixPathMatcher('custom/')

  it('matches paths longer than the prefix', () => {
    expect(matcher('custom/data.json')).toBe(true)
  })

  it('rejects bare prefix', () => {
    expect(matcher('custom/')).toBe(false)
  })

  it('rejects shorter-than-prefix', () => {
    expect(matcher('custom')).toBe(false)
  })

  it('rejects empty path', () => {
    expect(matcher('')).toBe(false)
  })
})

const delay = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms))
